### PR TITLE
Fix missing data from CSV download

### PR DESCRIPTION
**Pull request by echaidemenos**
resolves https://github.com/aqualinkorg/aqualink-app/issues/879

Basically moves CSV creation to the backend, fixing the issue of missing date when selecting `all dates`.
The issue was caused because the backend limits time-series data to 3 months if no other limit is specified.

### CODE REVIEW

File: packages/api/src/time-series/time-series.controller.ts [hunk 1]
```diff
@@ -173,4 +173,37 @@ export class TimeSeriesController {
...
+      'metrics',
+      new DefaultValuePipe(Object.values(Metric)),
+      ParseArrayPipe,
+    )
+    metrics: Metric[],
+    @Query('start', ParseDatePipe) startDate?: string,
+    @Query('end', ParseDatePipe) endDate?: string,
+    @Query('hourly', ParseBoolPipe) hourly?: boolean,
+  ) {
+    return this.timeSeriesService.findSiteDataCsv(
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/time-series/time-series.controller.ts:201]: I believe we have validation for the parameters. about dates we apply this validation pipe: https://github.com/aqualinkorg/aqualink-app/blob/master/packages/api/src/pipes/parse-date.pipe.ts
and about `siteId` for exaple we use 
```
@IsNumber()
 @Validate(EntityExists, [Site])
```
from `class-validator`

please correct me if i am missing something

File: packages/api/src/utils/time-series.utils.ts [hunk 2]
```diff
@@ -78,16 +78,30 @@ export const groupByMetricAndSource = <T extends TimeSeriesGroupable>(
...
+  hourly,
+  surveyPointId,
+  csv = false,
+}: GetDataQueryParams): Promise<TimeSeriesData[]> => {
+  const { endDate, startDate } = csv
+    ? { startDate: start, endDate: end }
+    : getTimeSeriesDefaultDates(start, end);
 
   const surveyPointCondition = surveyPointId
     ? `(source.survey_point_id = ${surveyPointId} OR source.survey_point_id is NULL)`
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/utils/time-series.utils.ts:107]: correct if i am wrong but we validate this by checking it is a number and that the survey point exists for that id:
```
@ApiProperty({ example: 1 })
@Type(() => Number)
@IsNumber()
@Validate(EntityExists, [SiteSurveyPoint])
surveyPointId: number;
```
here for example: https://github.com/aqualinkorg/aqualink-app/blob/master/packages/api/src/time-series/dto/survey-point-data-range.dto.ts

File: packages/website/src/common/Chart/MultipleSensorsCharts/DownloadCSVButton.tsx [hunk 3]
```diff
@@ -103,33 +45,24 @@ function DownloadCSVButton({
...
     }
 
-    if (!additionalData && !allDates && hourly) {
-      downloadCsv(getCSVData(data), undefined, fileName);
-      setOpen(false);
-      return;
-    }
+    const fileName = `data_site_${siteId}_${moment(startDate).format(
+      DATE_FORMAT,
+    )}_${moment(endDate).format(DATE_FORMAT)}.csv`;
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/website/src/common/Chart/MultipleSensorsCharts/DownloadCSVButton.tsx:50]: Thanks or pointing that out. We are opening an issue to resolve the problem!

File: packages/api/src/time-series/time-series.controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericboucher) [reply packages/api/src/time-series/time-series.controller.ts:201]: Graham can you explicit what you mean? The input validation already check that we are getting a valid date and a Boolean. The function will not get called if the validation doesn't pass.

File: packages/api/src/time-series/time-series.service.ts [hunk 4]
```diff
@@ -10,6 +10,8 @@ import {
   NotFoundException,
 } from '@nestjs/common';
 import { join } from 'path';
+// eslint-disable-next-line import/no-unresolved
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/time-series/time-series.service.ts:13]: It looks like yes or now. I will be linking this issue as comment also 
https://github.com/adaltas/node-csv/issues/372

[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/time-series/time-series.service.ts:13]: It looks like yes, for now, I will be adding a comment with a link to this issue
https://github.com/adaltas/node-csv/issues/372

File: packages/api/src/time-series/time-series.controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/time-series/time-series.controller.ts:201]: Ok i created a validation pipe for the metrics array. @ericboucher what do you think we should do about the dates? we still want i think some users to get all data they want? maybe restrict access to site admins?

[the remark this answers is not in the document]
Comment (ericboucher) [reply packages/api/src/time-series/time-series.controller.ts:210]: I think it's fine for now,
- I would add the start_date < end_date check
- And maybe there is a way to add a timeout to the query to limit risks?

File: packages/api/src/pipes/parse-metric-array.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  async transform(
+    value: string,
+    metadata: ArgumentMetadata,
+  ): Promise<string[]> {
+    if (!value) {
+      return this.options.defaultArray;
+    }
+
+    const parseArrayPipe = new ParseArrayPipe({
+      items: String,
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/pipes/parse-metric-array.pipe.ts:24]: `ParseArrayOptions` expect a type in `items` field: `items?: Type<unknown>;`

File: packages/api/src/pipes/parse-metric-array.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  ParseArrayPipe,
+  ArgumentMetadata,
+} from '@nestjs/common';
+
+@Injectable()
+export class MetricArrayPipe implements PipeTransform<any, Promise<string[]>> {
+  constructor(private readonly options: UniqueSubsetArrayOptions) {}
+
+  async transform(value: any, metadata: ArgumentMetadata): Promise<string[]> {
+    if (!value) {
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/pipes/parse-metric-array.pipe.ts:14]: We don't need to validate it is string. `transform` from `ParseArrayPipe` of `@nestjs/common`, used to simply parse arrays,  accepts `any`

File: packages/api/src/time-series/time-series.controller.ts [hunk 7]
```diff
@@ -171,6 +175,53 @@ export class TimeSeriesController {
     const file = this.timeSeriesService.getSampleUploadFiles(
       surveyPointDataRangeDto,
     );
+    res.set({
+      'Content-Disposition': `attachment; filename=${encodeURIComponent(
+        `${surveyPointDataRangeDto.source}_example.csv`,
```
[the remark this answers is not in the document]
Comment (echaidemenos) [reply packages/api/src/time-series/time-series.controller.ts:180]: it is the same:
```
res.set({
      'Content-Disposition': `attachment; filename=${encodeURIComponent(
        `${surveyPointDataRangeDto.source}_example.csv`,
      )}`,
    });
```
`encodeURIComponent` wraps 
```
`${surveyPointDataRangeDto.source}_example.csv`
```

**Review by ericboucher — APPROVED** [review]

**Pull request merged**